lib/luacrypto_aead.c

```diff
@@ -0,0 +1,325 @@
+	memcpy(buf, combined, combined_len); /* Copies AAD || CiphertextWithTag */
+	luaL_addsize(&b, combined_len);
+
+	/* Scatterlist covers AAD and CiphertextWithTag. */
+	struct scatterlist sg_work;
+	sg_init_one(&sg_work, buf, aad_len + ciphertext_with_tag_len);
+
+	struct aead_request *req = luacrypto_aead_prepare_request(L, tfm, aad_len, ciphertext_with_tag_len, &sg_work, iv_data);
+	if (!req) {
+		lunatik_free(iv_data);
+		lua_error(L);
+	}
+
+	int ret = crypto_aead_decrypt(req);
+	if (ret < 0) {
+		lunatik_free(iv_data);
+		luaL_error(L, "AEAD decryption failed (err %d)", ret);
+	}
+	aead_request_free(req);
+
+	/*
+	* After decryption, buf contains: AAD || Plaintext.
+	* The length of Plaintext is (ciphertext_with_tag_len - authsize_val)
+	*/
+	luaL_argcheck(L, ciphertext_with_tag_len >= authsize_val, 3,
+		"input data (ciphertext+tag) too short for tag"
+	);
+	/* The length of this combined data is (aad_len + plaintext_len) */
+	luaL_addsize(&b, -(int)authsize_val); /* Remove the tag length from the buffer size */
+	luaL_pushresult(&b);
+	lunatik_free(iv_data);
+	return 1;
+}
+
+/*** Lua C methods for the AEAD object.
+* Includes cryptographic operations and Lunatik metamethods.
+* The `__close` method is important for explicit resource cleanup.
+* @see crypto_aead
+* @see lunatik_closeobject
+*/
+static const luaL_Reg luacrypto_aead_mt[] = {
+	{"setkey", luacrypto_aead_setkey},
+	{"setauthsize", luacrypto_aead_setauthsize},
+	{"ivsize", luacrypto_aead_ivsize},
+	{"authsize", luacrypto_aead_authsize},
+	{"encrypt", luacrypto_aead_encrypt},
+	{"decrypt", luacrypto_aead_decrypt},
+	{"__gc", lunatik_deleteobject},
+	{"__close", lunatik_closeobject},
+	{"__index", lunatik_monitorobject},
+	{NULL, NULL}
+};
+
+/*** Lunatik class definition for AEAD TFM objects.
+* This structure binds the C implementation (luacrypto_aead_tfm_t, methods, release function)
+* to the Lua object system managed by Lunatik.
+*/
+static const lunatik_class_t luacrypto_aead_class = {
+	.name = "crypto_aead",
+	.methods = luacrypto_aead_mt,
+	.release = luacrypto_aead_release,
+	.sleep = true,
+	.pointer = true,
+};
+
+/*** Creates a new AEAD object.
+* This is the constructor function for the `crypto_aead` module.
+* @function crypto_aead.new
+* @tparam string algname The name of the AEAD algorithm (e.g., "gcm(aes)", "ccm(aes)").
+* @treturn crypto_aead The new AEAD object.
+* @raise Error if the TFM object or kernel request cannot be allocated/initialized.
+* @usage
+*   local aead = require("crypto_aead")
+*   local cipher = aead.new("gcm(aes)")
+*/
+static int luacrypto_aead_new(lua_State *L) {
+	const char *algname = luaL_checkstring(L, 1);
+	lunatik_object_t *object = lunatik_newobject(L, &luacrypto_aead_class, 0);
+
+	struct crypto_aead *tfm = crypto_alloc_aead(algname, 0, 0);
+	if (IS_ERR(tfm)) {
+		long err = PTR_ERR(tfm);
+		luaL_error(L, "Failed to allocate AEAD transform for %s (err %ld)", algname, err);
+	}
+	object->private = tfm;
+
+	return 1;
+}
```

Comment:
I would take a look on the constructors of the other crypto modules.. perhaps we can make a generic as well..